### PR TITLE
Added mocking in test_goes_ud.py to mock online tests

### LINKED ISSUE #2874: Work towards mocking all the online tests in SunPy

### PR DESCRIPTION
**Pull request by yashrsharma44**
1. Used mocking to simulate online test
2. Online test function to cross-check the mocking function

Partial work for #2874

### CODE REVIEW

File: sunpy/net/dataretriever/tests/test_goes_ud.py [hunk 1]
```diff
@@ -82,47 +149,79 @@ def test_query(LCClient, time):
...
-@pytest.mark.parametrize("time, instrument", [
-    (Time('1983/06/17', '1983/06/18'), Instrument('XRS')),
-    (Time('2012/10/4', '2012/10/6'), Instrument('XRS')),
-])
-def test_get(LCClient, time, instrument):
-    qr1 = LCClient.search(time, instrument)
+@mock.patch('sunpy.net.fido_factory.Fido.fetch',
+            side_effect=(UnifiedResponse(
+                mock_query_object('1983/06/17', '1983/06/18'))))
+@mock.patch('sunpy.net.download.Results.wait', return_value=['/home/yash/sunpy/data/go06830617.fits',
```
Comment (vn-ki) [inline sunpy/net/dataretriever/tests/test_goes_ud.py]: I'm pretty sure you didn't mean to include you're name :p

File: sunpy/net/dataretriever/tests/test_goes_ud.py [hunk 2]
```diff
@@ -15,9 +19,72 @@
 from sunpy.time import parse_time, is_time_equal
 
 
-@pytest.fixture
-def LCClient():
-    return goes.XRSClient()
+LCClient = goes.XRSClient()
+
+
+def mock_query_object(start_date, end_date):
```
Comment (vn-ki) [inline sunpy/net/dataretriever/tests/test_goes_ud.py]: Isn't this a lot of duplicated code from other PRs? Is there a better way to do this?

File: sunpy/net/dataretriever/tests/test_goes_ud.py [hunk 3]
```diff
@@ -15,9 +19,72 @@
...
+    # so no parsing required
+
+    sdate1 = mock_qr.time.start.value
+    sdate2 = datetime.strptime(qr.time.start.value, "%Y-%m-%dT%H:%M:%S.%f")
+
+    edate1 = mock_qr.time.end.value
+    edate2 = datetime.strptime(qr.time.end.value, "%Y-%m-%dT%H:%M:%S.%f")
+
+    assert abs(sdate2 - sdate1) < timedelta(seconds=1)
+    assert abs(edate2 - edate1) < timedelta(seconds=1)
```
Comment (vn-ki) [inline sunpy/net/dataretriever/tests/test_goes_ud.py]: You should be using `sunpy.time.astropy_time._is_time_equal` to check the equality of the `Time` objects.

Comment (yashrsharma44) [reply sunpy/net/dataretriever/tests/test_goes_ud.py]: Shouldn't we refrain from using private methods?

Comment (vn-ki) [reply sunpy/net/dataretriever/tests/test_goes_ud.py]: This function is made for use in tests exclusively. There might be rounding errors when comparing astropy times.

File: sunpy/net/dataretriever/tests/test_goes_ud.py (hunk 2 above)
Comment (yashrsharma44) [reply sunpy/net/dataretriever/tests/test_goes_ud.py]: I know that this is duplicated, but there are few steps which require different processing :smile:

File: sunpy/net/dataretriever/tests/test_goes_ud.py (hunk 3 above)
Comment (yashrsharma44) [reply sunpy/net/dataretriever/tests/test_goes_ud.py]: Okay, that should help me replace a crude method.

Comment (nabobalis) [reply sunpy/net/dataretriever/tests/test_goes_ud.py]: Also there should be nothing wrong with using private sunpy functions. It is using upstream ones that might  be a problem down the road.

Comment (yashrsharma44) [reply sunpy/net/dataretriever/tests/test_goes_ud.py]: Are you sure that this method exists, for I am getting this error - `ModuleNotFoundError: No module named 'sunpy.time.astropy_time'
`

Comment (nabobalis) [reply sunpy/net/dataretriever/tests/test_goes_ud.py]: The function was moved into `sunpy.time.time` if I recall and the private name was dropped.

Comment (yashrsharma44) [reply sunpy/net/dataretriever/tests/test_goes_ud.py]: Okay, so I should stick to my original convention.

Comment (vn-ki) [reply sunpy/net/dataretriever/tests/test_goes_ud.py]: > The function was moved into sunpy.time.time if I recall and the private name was dropped.

0_0

File: sunpy/net/dataretriever/tests/test_goes_ud.py [hunk 4]
```diff
@@ -15,9 +19,72 @@
...
+        'Time_end':  parse_time(end_date),
+        'source': 'nasa',
+        'instrument': 'goes',
+        'physobs': 'irradiance',
+        'provider': 'sdac'
+    }
+
+    st_datetime = datetime.strptime(start_date, '%Y/%m/%d')
+    ed_datetime = datetime.strptime(end_date, '%Y/%m/%d') + timedelta(days=1)
+    time_range = TimeRange(st_datetime, ed_datetime).split(int((ed_datetime-st_datetime).days))
```
Comment (vn-ki) [inline sunpy/net/dataretriever/tests/test_goes_ud.py]: Isn't this the same as
```py
tr = map_['TimeRange']
time_range = tr.split(int((tr.days + 1).value))
```

File: sunpy/net/dataretriever/tests/test_goes_ud.py [hunk 5]
```diff
@@ -9,15 +13,81 @@
...
+    # so no parsing required
+
+    sdate1 = mock_qr.time.start.value
+    sdate2 = datetime.strptime(qr.time.start.value, "%Y-%m-%dT%H:%M:%S.%f")
+
+    edate1 = mock_qr.time.end.value
+    edate2 = datetime.strptime(qr.time.end.value, "%Y-%m-%dT%H:%M:%S.%f")
+
+    # Use _is_time_equal to check if the dates are almost equal
+    assert is_time_equal(sdate2,sdate1)
```
Comment (vn-ki) [inline sunpy/net/dataretriever/tests/test_goes_ud.py]: Is `mock_qr.time` and `qr.time`, `TimeRange` objects? If yes, you can do equality check directly on them.

Else you can check start of both with `is_time_equal`. Don't need conversion to datetime here.

File: sunpy/net/dataretriever/tests/test_goes_ud.py (hunk 3 above)
Comment (yashrsharma44) [reply sunpy/net/dataretriever/tests/test_goes_ud.py]: I found out that `TimeRange` cannot be subtracted, so I am going to drop this idea.

File: sunpy/net/dataretriever/tests/test_goes_ud.py (hunk 4 above)
Comment (yashrsharma44) [reply sunpy/net/dataretriever/tests/test_goes_ud.py]: The `TimeRange` object is getting split into parts, which is not what I intend. So I am dropping this suggestion

File: sunpy/net/dataretriever/tests/test_goes_ud.py (hunk 5 above)
Comment (yashrsharma44) [reply sunpy/net/dataretriever/tests/test_goes_ud.py]: `is_time_equal` does not work for `TimeRange` objects.

**Pull request closed without merging**